README.md
ZEsarUX - ZX Second-Emulator And Released for UniX 

Created by Cesar Hernandez Bano

https://github.com/chernandezba/zesarux


It's a ZX Machines Emulator for UNIX based operating systems (and Windows), including all the Sinclair computers:

* MK14
* ZX80
* ZX81
* ZX Spectrum
* QL
* Z88

And also:

* Timex TS 2068
* Sam Coupe
* Pentagon 
* Chloe 140 SE, 280 SE
* Chrome
* Prism
* ZX-Uno
* ZX-Evolution BaseConf
* ZX-Evolution TS-Conf
* TBBlue/ZX Spectrum Next
* Jupiter Ace
* Amstrad CPC 464, CPC 4128

ZEsarUX source code and binaries are distributed under GNU GPL license. 
ZEsarUX also includes a folder, "my_soft", which has some programs and data made by me. The entire folder is also covered by the GNU GPL license.

ZEsarUX also includes third-party roms, media, programs and games NOT covered by this license.

This is my second ZX Spectrum emulator after ZXSpectr
https://github.com/chernandezba/zxspectr

I recommend you to read FEATURES, INSTALL and HISTORY files, as well as other documents in this emulator.
You can open them from the help menu or from an external viewer.


ZEsarUX distributed under GNU GENERAL PUBLIC LICENSE v3. You may read it on the LICENSE file.

Please read the other licenses used in ZEsarUX, from the menu Help->Licenses or just open files from folder licenses/


Available versions for download here:
* Source code
* Binary compiled versions:
* GNU/Linux 32/64 Bits
* FreeBSD 64 Bits
* Mac OS X
* Windows
* Raspberry pi (raspbian)

Other by 3rd party:
* Arch Linux
* Slackware
* Retropie/EmulationStation
* Open Pandora
* PocketCHIP
* MorhpOS


ZEsarUX has won the "Best Emulator" award from Retrogaming Total blog on 2015 and 2017




__DONATE__

ZEsarUX is free software and you don't need to pay to use it. 
ZEsarUX will always cost you nothing to use, but that doesn't mean it costs me nothing to make.
So if you want to demonstrate your appreciation to it, you can donate using Paypal. Just click:

[ZEsarUX donation](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=E5RSRST8N7KWS&lc=ES&item_name=Donativo%20ZEsarUX&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donateCC_LG%2egif%3aNonHosted)


All donors will appear in the acknowledgements file

Thanks!


__FEATURES__

* Runs on any Unix system and Windows: tested on GNU/Linux x86, GNU/Linux x86_64, GNU/Linux Raspbian, GNU/Linux RetroPie, FreeBSD x86_64, Mac OS X, Windows native, Windows + Cygwin

* Emulates:
* Science of Cambridge MK14
* Sinclair ZX80
* Sinclair ZX81
* Spectrum models: 16k, 48k (English and Spanish), Inves Spectrum +, Spectrum 128k (English and Spanish), Spectrum +2 (English, Spanish and French), Spectrum +2A (English and Spanish), Spectrum +3 (English and Spanish), Microdigital TK90X (Portuguese and Spanish), TK95 
* Sinclair QL
* Cambridge Z88
* Timex TS 2068 
* Sam Coupe
* Pentagon 
* Chloe 140 SE, 280 SE
* Chrome
* Prism
* ZX-Uno
* ZX-Evolution BaseConf (beta)
* ZX-Evolution TS-Conf
* TBBlue/ZX Spectrum Next
* Jupiter Ace
* Amstrad CPC 464, CPC 4128

* Is the only emulator for machines: Chloe 140/280, Prism and Chrome

* Almost perfect emulation of timing of all Spectrum machines

* Supports undocumented Z80 flags, features, and MEMPTR register

* Supports idle bus port, contended memory, ULA early/late timings

* Supports Motorola CPU 68008 (Sinclair QL)

* Supports SC/MP CPU - INS8060 (MK14)

* Supports perfect ZX Spectrum 16k/48k colour palette

* Supports ULAplus: Standard 64 colour palette, linear modes with 16 colours per pixel: Radastan Mode: 128x96, ZEsarUX modes: 256x96, 128x192, 256x192

* Supports Chroma81 ZX81 Colour mode

* Supports Spectra Video Modes

* Supports Timex Video modes (Mode 0 standard, 1 dual screen, 2 hires colour 8x1, 6 512x192 monochrome)

* Supports Pentagon 16C mode

* Supports All Video modes from Prism machine

* Supports ZGX Sprite Chip

* Supports snow effect on Spectrum models

* Supports interlaced, scanlines and Gigascreen effects

* Supports hi-res modes on ZX Spectrum (rainbow effects and others) and ZX80/ZX81 (UDG, CHR$128, WRX, HRG and some other hi-res modes)

* Supports reducing the screen to 192x144 (0.75 scale)

* Uses its own powerful window environment (ZX Vision) inside the program application window, having the same GUI style on all plattforms

* Can be used with joystick and environments without keyboard, like Raspiboy / Retropie

* Emulates all the oddities of the Inves Spectrum +: 64 KB RAM, RAM initialization with FF00H pattern, OUT ula AND RAM, EAR and MIC XOR, no contended memory, snow in border, colour ula delay, interrupt starts at end of top border (not at the beginning of the border), corrupt memory on every interrupt, no idle bus

* Uses Video drivers: X-Windows, SDL, Cocoa (Mac OS X), Framebuffer, ncurses, aalib, cacalib, stdout(console), simpletext(console)

* Uses Audio Drivers: PulseAudio, Alsa, SDL, DSP(OSS), CoreAudio (Mac OS X). 

* Supports dumping audio & video to file

* Supports real tape loading of file formats: RWA, SMP, WAV, TZX, PZX, CDT, TAP, P, O. It handles loading of turbo load tapes or any type of tape loading for standard/non standard loading routines

* Handles binary tape format files (TAP, TZX, O, P, CDT) on standard BASIC ROM routines for ZX Spectrum, ZX80, ZX81, Jupiter Ace

* Handles real audio loading (RWA, SMP, WAV) on standard BASIC ROM routines for ZX Spectrum, ZX80, ZX81, Amstrad CPC, Jupiter Ace

* Simulates real tape loading on standard BASIC ROM routines for ZX Spectrum, ZX80, ZX81

* Handles snapshot format: ZSF, ZX, Z80, SP, SPG, NEX, SNA, P, O, Z81, ACE

* Handles RZX playback 

* Emulates Multiface One, 128 and Three

* Emulates Datagear/MB02 DMA

* Handles DCK Timex dock cartridges

* Emulates +3 Disk

* Emulates Betadisk/TR-DOS. Direct support for .trd files. scl files can be converted from file selector pressing space

* Emulates ZX Spectrum MMC Interfaces: ZXMMC, DivMMC. Support for raw (.mmc) files and hdf files

* Emulates ZX Spectrum IDE Interface: DivIDE, 8-bit simple 

* Emulates Sam Coupe IDE Interface: Atom Lite

* Emulates ZXPand MMC Interface on ZX80 and ZX81

* Emulates ZX Dandanator! Mini, CPC Dandanator! Mini

* Emulates Speccy Superupgrade

* Emulates Kartusho,iFrom interfaces by Antonio Villena

* Emulates DataDrive HiLow (beta)

* Handles RAM, EPROM, Intel Flash and Hybrid (RAM+EPROM) cards on Z88

* Handles compressed formats zip (with internal decompressor), and tar, rar, gz through external utilities

* Supports AY Chip, Turbosound (2 AY Channels), 3 AY Channels, MIDI channels, different DAC: Specdrum, Covox, ACB/ABC/BAC Stereo 

* Real time playback from AY Chip sound to external MIDI device 

* Supports Quicksilva, ZON-X81, and VSYNC-based sound on ZX80/81

* Supports Speaker and Soundbox sound on Jupiter Ace

* Experimental simulation of the Sam Coupe Audio Chip (SAA1099)

* Supports RAM size up to 1024 KB on Spectrum and Pentagon

* Supports RAM packs on ZX80/81 up to 56 KB

* Supports RAM packs on Jupiter Ace up to 35 KB

* Joystick emulation with real joystick (on GNU/Linux and Windows) and keyboard cursors: Kempston, Sinclair 1&2, Cursor Joystick, Cursor & Shift, OPQA, Fuller, Zebra, Mikro-Gen, ZXPand, SAM Coupe Cursors. Autofire function too

* On Screen keyboard useful when playing with joystick, two types: one with keyboard letters, and another with words, useful for playing Text Adventures. Also a tool to extract words from text adventures (Daad, Paws, Quill and Gac) and configure this on screen text keyboard

* Emulates Spectrum keyboard ghosting error feature

* ZX Spectrum Recreated keyboard support

* Supports native turbo modes on ZX-Uno, Chloe, Prism and TBBlue and manual for other machines

* ZX Printer emulation

* Lightgun emulation: Almost perfect emulation of Gunstick from MHT Ingenieros S.L and experimental emulation of Magnum Light Phaser

* Kempston mouse emulation

* Supports Network gaming using its own protocol (ZEsarUX Network Gaming protocol - ZENG), which allows you to play to any emulated game, using two ZEsarUX instances, located each one on any part of the world or in a local network. Games doesn't have to be modified, you can use any existing game

* Uart bridge emulation: allows you to use a real uart-wifi connected to your computer, on GNU/Linux only (Windows, Mac can simulate it)

* Can browse online ZX81 games

* Can browse online Spectrum games

* Supports Input spool text file to simulate keyboard press

* Supports reading Pokes from .POK files

* AY Chip Tools: AY Player, AY Sheet, AY Piano, AY Registers, export AY music to midi (.mid) files, AY Mixer

* Supports ZEsarUX remote command protocol (ZRCP). You can do enhanced debugging on ZEsarUX by using a telnet client.

* Can use a reduced Spectrum core, with some features disabled, useful on slow devices, like Raspberry Pi 1/Zero

* Powerful debug features: Assembler, Registers, Dissassemble, Conditional Breakpoints using text expressions, Watches, Step-to-step, Step-over, Runto, Show TV electron position, Load source code, Hexadecimal Editor, View Sprites, Find byte (useful to find POKEs), CPU Transaction log, View BASIC, verbose messages on console, DAAD/PAWS/Quill Debugger

* File utilities menu: Allowing to view, expand and convert some common file formats: tap, tzx, trd, dsk, mdv, hdf, etc

* Accessibility support: Print char traps allows to capture generated text from almost any program or game, using standard ROM calls (RST 10H) or even non standard print character routines. Can send generated text from a game to a text-to-speech program. It's ready for text to speech support for blind or visually impaired people. Menu emulator can also be read by a text-to-speech program.

* Simulates upper RAM memory refresh on Spectrum 48Kb, losing its contents when changing R register very quickly

* Supports command line settings, configuration file settings and per-game configuration settings

* Other features: Visualmem menu, CPU Statistics

* Includes three easter eggs. Can you find them? :)



__Some screenshots__

ZX Spectrum Overscan demo


src/Changelog
Version 8.2 - xxx edition. 16 March 2020

Added support for compilation on FreeBSD
Added F key to toggle top cpu speed
Added emulation of AY chip MIDI registers 14 & 15 and route them to an external MIDI device
Added Visualmem for MMC Read/Write
Added Visualmem "Defrag style"
Added browser for quicksave snapshots
Added save screen to bmp feature
Added Solarized Dark & Light Gui Style

Added ZRCP commands:
-save-screen: to save machine screen to file



Improved menu:
-Allow some windows to be put on the background. Also active windows are saved and restored on startup
-Pressing the menu key (F5 by default) when menu is open, it will send you back to the main menu. Also pressing over Z logo on ext desktop
-Added setting to disable showing cpu temperature, or FPS on footer

Improved Esxdos handler: 
-add f_mkdir function 
-add configuration setting to save local work directory

Improved ZRCP:
-Allow to write commands in "". Not all commands support this
-Added save-binary command

Improved TBBlue emulation:
-Enable Timex video by default
-Added altrom feature: now 48k and 128k mode works again
-Legacy hi-res and border effects disabled by default. Reduces up to 13% cpu use or gain 17 FPS on my environment
-Emulate register 0x8E (142) => Spectrum 128K Memory Mapping
-Allow to download a 512 MB TBBlue SD official card (allowed download sizes: 32MB, 128MB, 512MB, 2 GB)
-Added Layer2 modes 320x256 and 640x256
-Allow paths with backslashes on esxdos handler

Improved debugging:
-Added breakpoint action "reset-tstatp"
-Added SEG0, SEG1, ... SEG7 variables to match TBBlue MMU pages

Improved Chloe emulation: 
-Support CPU speed setting via the ZX-Uno SCANDBLCTRL register
-Load unodos3.rom as the default divmmc firmware
-Added chloehd.mmc in the extras package

Improved emulation: use a slightly less cpu on every scanline

Fixed ZRCP bugs:
-Fix possible crash when disconnecting socket

Fixed rom loading: no more cpu panic when the rom is not found 

Fixed Esxdos handler: 
-fixed segfaults when using a file handler that was a directory

Fixed SCF/CCF undocumented flags emulation. Not enabled by default on compilation because almost no one is using this
Fixed saving "Show CPU usage" setting on footer


Fixed tbblue bugs:
-Fixed 8 first lines of tiles on border not shown
-Fixed text mode (1 bit tile)



Version 8.1.1 - The Great Escape edition. 2 March 2020

Fixed starting the emulator on some Mac OS versions


Version 8.1 - The Great Escape edition. 25 February 2020


Added ZEsarUX Network Gaming protocol (ZENG), which allows you to play to any emulated game, using two ZEsarUX instances, 
located each one on any part of the world or in a local network. Games doesn't have to be modified, you can use any existing game

Added Spectrum and ZX81 online game browser. You can search & play almost any game for these machines!
Added real joystick support on Windows
Added Pentagon 16C colour mode emulation
Added 1024 KB RAM emulation for Spectrum & Pentagon
Added ZIP files support: no more need to use external unzip program
Added SCL files support: can be converted from file selector menu pressing space
Added SNA saving support
Added PZX tape file support
Added SCR to TAP file converter

Added total run time statistics 
Added checking for updates
Added sending anonymous statistics use
Added setting --disablemenuandexit: it closes the emulator when the menu is going to be opened. 
 It's the same behaviour as previous --disablemenu. Now --disablemenu only disables the menu but does not close the emulator
Added setting to disable looking for alternate character sets from sysvar 23606/7 on ocr functions
Added DataDrive HiLow emulation (beta)
Added Borland Gui Style
Added setting to reset configuration file to default values
Added setting to choose cpu type

Added ZRCP commands:
-cpu-code-coverage: working with code coverage
-cpu-history: working with cpu history
-extended-stack: having an internal stack for debugging stack value types
-get-ui-io-ports: to get keyboard state and joystick
-get-snapshot: get a ZSF snapshot from ZRCP
-print-footer: print message on footer
-put-snapshot: put a ZSF snapshot from ZRCP
-send-keys-event: simulates key press/release
-set-ui-io-ports: to change keyboard state and joystick
-write-port: write value to a port

Improved TBBlue emulation:
-New Wizard to automatically download and configure TBBlue official SD images
-optimized ula,lores render routines: 2% less cpu
-added scroll x,y on ula layer
-added 1-bit tile mode
-added tbblue register 63H: copper 16 bit write 
-added tbblue register 69H: display control 1 register
-added tbblue registers 0x35-0x39, 0x75-0x79
-added copper opcodes NOOP, HALT
-increased tbblue.mmc MMC bundled image size to 64 MB
-updated to the latest tbblue SD distribution. 
-enable real video by default
-added zsf snapshot support
-supports setting Core ID to ZX Dos
-supports changing Core version

Improved Debugging: 
-added setting to dump .zsf snapshot when a cpu panic is fired

src/cpu.h
/*
    ZEsarUX  ZX Second-Emulator And Released for UniX
    Copyright (C) 2013 Cesar Hernandez Bano

    This file is part of ZEsarUX.

    ZEsarUX is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef CPU_H
#define CPU_H

#include "compileoptions.h"

/*
Emulator version format:
XX.YY-S

XX: Major version number. 1 or 2 digits number. Currently using only one digit but could be greater (on ZEsarUX version 10 or higher)
YY: Minor version number. 1 or 2 digits number
-S: Optional string, only present on non-stable versions. S it's normally "SN" or "RC", where RC can also be RCX, being X a digit: RC1, RC2, etc

Examples

6.0: Means major version 6, minor version 0. It's a stable version (no suffix string -....)

6.1-RC: Means major version 6, minor version 1, beta version: RC
6.2-RC3: Means major version 6, minor version 2, beta version: RC3

*/

//#define EMULATOR_VERSION "8.2"

#define EMULATOR_VERSION "8.2-Ped7g"
//#define EMULATOR_VERSION "8.2-RC"
#define SNAPSHOT_VERSION

#define EMULATOR_DATE "16 March 2020"
#define EMULATOR_SHORT_DATE "16/03/2020"
#define EMULATOR_GAME_EDITION "xxx"
#define EMULATOR_EDITION_NAME EMULATOR_GAME_EDITION " edition"

//8 bits
typedef unsigned char z80_byte;
//16 bits
typedef unsigned short z80_int;
//32 bits
typedef unsigned int z80_long_int;

//Usado en menu, para registros de 16 bits de z80 y de 32 de motorola
typedef unsigned int menu_z80_moto_int;

struct s_z80_bit {
	unsigned int v:1;
};


typedef union {
#ifdef WORDS_BIGENDIAN
  struct { z80_byte h,l; } b;
#else
  struct { z80_byte l,h; } b;
#endif

  z80_int w;
} z80_registro;


typedef struct s_z80_bit z80_bit;


extern int zoom_x,zoom_y;
extern int zoom_x_original,zoom_y_original;


extern z80_registro registro_hl;
#define reg_h registro_hl.b.h
#define reg_l registro_hl.b.l
#define HL registro_hl.w
#define reg_hl registro_hl.w


extern z80_registro registro_de;
#define reg_d registro_de.b.h
#define reg_e registro_de.b.l
#define DE registro_de.w
#define reg_de registro_de.w


extern z80_registro registro_bc;
#define reg_b registro_bc.b.h
#define reg_c registro_bc.b.l
#define BC registro_bc.w
#define reg_bc registro_bc.w


extern z80_byte reg_i;
extern z80_byte reg_r,reg_r_bit7;

#define IR ( ( reg_i ) << 8 | ( reg_r_bit7 & 0x80 ) | ( reg_r & 0x7f ) )

extern z80_byte reg_h_shadow,reg_l_shadow;
extern z80_byte reg_b_shadow,reg_c_shadow;
extern z80_byte reg_d_shadow,reg_e_shadow;
extern z80_byte reg_a_shadow;

#define REG_AF (value_8_to_16(reg_a,Z80_FLAGS))

#define REG_AF_SHADOW (value_8_to_16(reg_a_shadow,Z80_FLAGS_SHADOW))
#define REG_HL_SHADOW (value_8_to_16(reg_h_shadow,reg_l_shadow))
#define REG_BC_SHADOW (value_8_to_16(reg_b_shadow,reg_c_shadow))
#define REG_DE_SHADOW (value_8_to_16(reg_d_shadow,reg_e_shadow))

extern void set_machine(char *romfile);
extern void set_machine_params(void);
extern void post_set_machine(char *romfile);
extern void post_set_machine_no_rom_load(void);

extern z80_int reg_pc;
extern z80_byte reg_a;
extern z80_int reg_sp;
extern z80_int reg_ix;
extern z80_int reg_iy;

extern z80_byte Z80_FLAGS;
extern z80_byte Z80_FLAGS_SHADOW;

#define FLAG_C  0x01
#define FLAG_N  0x02
#define FLAG_PV 0x04
#define FLAG_3  0x08
#define FLAG_H  0x10
#define FLAG_5  0x20
#define FLAG_Z  0x40
#define FLAG_S  0x80

extern z80_int memptr;

extern z80_byte scf_ccf_undoc_flags_before;
extern int scf_ccf_undoc_flags_after_changed;

extern z80_bit iff1,iff2;

extern z80_bit interrupcion_pendiente;
extern z80_bit z80_ejecutando_halt;
extern z80_byte im_mode;
extern z80_bit cpu_step_mode;
extern int core_refetch;
extern int cpu_duracion_pulso_interrupcion;
extern z80_bit core_end_frame_check_zrcp_zeng_snap;

#ifndef GCC_UNUSED

#ifdef __GNUC__
#  define GCC_UNUSED __attribute__((unused))
#else
#  define GCC_UNUSED
#endif

#endif


extern z80_byte puerto_65278; //    db    255  ; V    C    X    Z    Sh    ;0
extern z80_byte puerto_65022; //    db    255  ; G    F    D    S    A     ;1
extern z80_byte puerto_64510; //    db              255  ; T    R    E    W    Q     ;2
extern z80_byte puerto_63486; //    db              255  ; 5    4    3    2    1     ;3
extern z80_byte puerto_61438; //    db              255  ; 6    7    8    9    0     ;4
extern z80_byte puerto_57342; //    db              255  ; Y    U    I    O    P     ;5
extern z80_byte puerto_49150; //    db              255  ; H                J         K      L    Enter ;6
extern z80_byte puerto_32766; //    db              255  ; B    N    M    Simb Space ;7
extern z80_byte puerto_especial1;
extern z80_byte puerto_especial2;
extern z80_byte puerto_especial3;
extern z80_byte puerto_especial4;


extern z80_int *registro_ixiy;

extern z80_bit border_enabled;


extern void (*codsinpr[]) ();
extern void (*codprcb[]) ();
extern void (*codpred[]) ();

extern void (*codprddfd[]) ();

extern z80_byte *memoria_spectrum;

//extern z80_byte inves_ula_delay_factor;
extern z80_byte current_machine_type;
extern z80_bit modificado_border;

extern z80_bit inves_ula_bright_error;




extern void set_undocumented_flags_bits(z80_byte value);
extern void set_flags_zero_sign(z80_byte value);
extern void set_flags_zero_sign_16(z80_int value);
extern void set_flags_carry_suma(z80_byte antes,z80_byte result);
extern void set_flags_carry_resta(z80_byte antes,z80_byte result);
extern void set_flags_carry_16_suma(z80_int antes,z80_int result);
extern void set_flags_carry_16_resta(z80_int antes,z80_int result);

extern void reset_cpu(void);
extern void cold_start_cpu_registers(void);

#define value_8_to_16(h,l) (((h)<<8)|l)

#define value_16_to_8l(hl) ((hl) & 0xFF)
#define value_16_to_8h(hl) (((hl)>>8) & 0xFF)

extern z80_byte *devuelve_reg_offset(z80_byte valor);

extern int video_zx8081_linecntr;
extern z80_bit video_zx8081_linecntr_enabled;

extern z80_byte video_zx8081_ula_video_output;


//#define TEMP_LINEAS_ZX8081 192
//#define TEMP_BORDE_SUP_ZX8081 56


//#define macro_invalid_opcode(MESSAGE) debug_printf(VERBOSE_ERR,"Invalid opcode " MESSAGE ". Final PC: %X",reg_pc)

//extern z80_bit debug_cpu_core_loop;
//extern int cpu_core_loop_active;
//extern void (*cpu_core_loop) (void);

#define CPU_CORE_SPECTRUM 1
#define CPU_CORE_ZX8081 2
#define CPU_CORE_Z88 3
#define CPU_CORE_ACE 4
#define CPU_CORE_CPC 5
#define CPU_CORE_SAM 6
#define CPU_CORE_QL 7
#define CPU_CORE_MK14 8

extern struct timeval z80_interrupts_timer_antes, z80_interrupts_timer_ahora;
extern long z80_timer_difftime, z80_timer_seconds, z80_timer_useconds;
extern struct timeval zesarux_start_time;
extern z80_bit interrupcion_timer_generada;
extern int contador_frames_veces_buffer_audio;
extern z80_bit esperando_tiempo_final_t_estados;

extern z80_bit interrupcion_maskable_generada;
extern z80_bit interrupcion_non_maskable_generada;
extern z80_bit interrupcion_timer_generada;
extern z80_byte reg_r_antes_zx8081;
extern z80_bit temp_zx8081_lineasparimpar;
extern char *get_machine_name(z80_byte m);

extern z80_bit cpu_random_r_register;

extern int porcentaje_velocidad_emulador;
extern void set_emulator_speed(void);

//T-estados totales del frame
extern int t_estados;

extern int t_scanline;

//Scan line actual
extern int t_scanline_draw;
extern int t_scanline_draw_timeout;

//Autoseleccionar opciones de emulacion (audio, realvideo, etc) segun snap o cinta cargada
extern z80_bit autoselect_snaptape_options;

extern z80_bit tape_loading_simulate;
extern z80_bit tape_loading_simulate_fast;
extern void end_emulator(void);
extern void end_emulator_saveornot_config(int saveconfig);

extern z80_bit snow_effect_enabled;

struct s_driver_struct
{
        char driver_name[30];
        int (*funcion_init) () ;
        int (*funcion_set) () ;
};

typedef struct s_driver_struct driver_struct;

extern driver_struct scr_driver_array[];
extern int num_scr_driver_array;

extern driver_struct audio_driver_array[];
extern int num_audio_driver_array;

#define MAX_SCR_INIT 10
#define MAX_AUDIO_INIT 10

extern z80_bit stdout_simpletext_automatic_redraw;
//Valores para stdout. Estan aqui porque se graban en archivo .zx, aunque no este el driver stdout compilado
//fin valores para stdout


#define MACHINE_ID_SPECTRUM_16			0
#define MACHINE_ID_SPECTRUM_48			1
#define MACHINE_ID_INVES			2
#define MACHINE_ID_MICRODIGITAL_TK90X		3
#define MACHINE_ID_MICRODIGITAL_TK90X_SPA	4
#define MACHINE_ID_MICRODIGITAL_TK95		5
#define MACHINE_ID_SPECTRUM_128			6
#define MACHINE_ID_SPECTRUM_128_SPA		7
#define MACHINE_ID_SPECTRUM_P2			8
#define MACHINE_ID_SPECTRUM_P2_FRE		9
#define MACHINE_ID_SPECTRUM_P2_SPA		10
#define MACHINE_ID_SPECTRUM_P2A_40		11
#define MACHINE_ID_SPECTRUM_P2A_41		12
#define MACHINE_ID_SPECTRUM_P2A_SPA		13
#define MACHINE_ID_ZXUNO			14
#define MACHINE_ID_CHLOE_140SE			15
#define MACHINE_ID_CHLOE_280SE			16
#define MACHINE_ID_TIMEX_TS2068			17
#define MACHINE_ID_PRISM			18
#define MACHINE_ID_TBBLUE			19
#define MACHINE_ID_SPECTRUM_48_SPA		20

#define MACHINE_ID_PENTAGON			21
#define MACHINE_ID_CHROME			22

#define MACHINE_ID_TSCONF			23
#define MACHINE_ID_BASECONF			24


#define MACHINE_ID_SPECTRUM_P3_40		25
#define MACHINE_ID_SPECTRUM_P3_41		26
#define MACHINE_ID_SPECTRUM_P3_SPA		27


#define MACHINE_ID_ZX80				120
#define MACHINE_ID_ZX81				121
#define MACHINE_ID_ACE				122
#define MACHINE_ID_Z88				130

#define MACHINE_ID_CPC_464			140
#define MACHINE_ID_CPC_4128			141
#define MACHINE_ID_SAM				150
#define MACHINE_ID_QL_STANDARD			160
#define MACHINE_ID_MK14_STANDARD		180



//Condiciones de maquinas activas
#define MACHINE_IS_SPECTRUM (current_machine_type<30)
#define MACHINE_IS_SPECTRUM_16 (current_machine_type==MACHINE_ID_SPECTRUM_16)
#define MACHINE_IS_SPECTRUM_48 (current_machine_type==MACHINE_ID_SPECTRUM_48)
#define MACHINE_IS_INVES (current_machine_type==MACHINE_ID_INVES)
#define MACHINE_IS_SPECTRUM_48_SPA (current_machine_type==MACHINE_ID_SPECTRUM_48_SPA)
#define MACHINE_IS_SPECTRUM_16_48 ( (current_machine_type<=MACHINE_ID_MICRODIGITAL_TK95) || MACHINE_IS_SPECTRUM_48_SPA )

#define MACHINE_IS_SPECTRUM_128_P2 ( (current_machine_type>=MACHINE_ID_SPECTRUM_128 && current_machine_type<=MACHINE_ID_SPECTRUM_P2_SPA) || MACHINE_IS_PENTAGON)

#define MACHINE_IS_SPECTRUM_P3 (current_machine_type==MACHINE_ID_SPECTRUM_P3_40 || current_machine_type==MACHINE_ID_SPECTRUM_P3_41 || current_machine_type==MACHINE_ID_SPECTRUM_P3_SPA)

//MACHINE_IS_SPECTRUM_P2A_P3 antes de emular el +3, era MACHINE_IS_SPECTRUM_P2A
#define MACHINE_IS_SPECTRUM_P2A_P3 ( (current_machine_type>=MACHINE_ID_SPECTRUM_P2A_40 && current_machine_type<=MACHINE_ID_SPECTRUM_P2A_SPA) || MACHINE_IS_SPECTRUM_P3)

//MACHINE_IS_SPECTRUM_128_P2_P2A_P3 era MACHINE_IS_SPECTRUM_128_P2_P2A antes de emular el +3
#define MACHINE_IS_SPECTRUM_128_P2_P2A_P3 ( MACHINE_IS_SPECTRUM_128_P2 || MACHINE_IS_SPECTRUM_P2A_P3)

#define MACHINE_IS_ZXUNO (current_machine_type==MACHINE_ID_ZXUNO)

#define MACHINE_IS_CHLOE_140SE (current_machine_type==MACHINE_ID_CHLOE_140SE)
#define MACHINE_IS_CHLOE_280SE (current_machine_type==MACHINE_ID_CHLOE_280SE)
#define MACHINE_IS_CHLOE (MACHINE_IS_CHLOE_140SE || MACHINE_IS_CHLOE_280SE)
#define MACHINE_IS_TIMEX_TS2068 (current_machine_type==MACHINE_ID_TIMEX_TS2068)
#define MACHINE_IS_PRISM (current_machine_type==MACHINE_ID_PRISM)
#define MACHINE_IS_TBBLUE (current_machine_type==MACHINE_ID_TBBLUE)
#define MACHINE_IS_PENTAGON (current_machine_type==MACHINE_ID_PENTAGON)
#define MACHINE_IS_CHROME (current_machine_type==MACHINE_ID_CHROME)


#define MACHINE_IS_TSCONF (current_machine_type==MACHINE_ID_TSCONF)
#define MACHINE_IS_BASECONF (current_machine_type==MACHINE_ID_BASECONF)

#define MACHINE_IS_ZXEVO (MACHINE_IS_TSCONF || MACHINE_IS_BASECONF)

#define MACHINE_IS_ZX80 (current_machine_type==MACHINE_ID_ZX80)
#define MACHINE_IS_ZX81 (current_machine_type==MACHINE_ID_ZX81)
#define MACHINE_IS_ZX8081 (MACHINE_IS_ZX80 || MACHINE_IS_ZX81)
#define MACHINE_IS_ACE (current_machine_type==MACHINE_ID_ACE)
#define MACHINE_IS_ZX8081ACE (MACHINE_IS_ZX8081 || MACHINE_IS_ACE)

#define MACHINE_IS_Z88 (current_machine_type==MACHINE_ID_Z88)

#define MACHINE_IS_CPC_464 (current_machine_type==MACHINE_ID_CPC_464)
#define MACHINE_IS_CPC_4128 (current_machine_type==MACHINE_ID_CPC_4128)
#define MACHINE_IS_CPC (current_machine_type>=MACHINE_ID_CPC_464 && current_machine_type<=149)
#define MACHINE_IS_SAM (current_machine_type==MACHINE_ID_SAM)

/*
160=QL
161-179 reservado para otros QL
*/

#define MACHINE_IS_QL_STANDARD (current_machine_type==MACHINE_ID_QL_STANDARD)
#define MACHINE_IS_QL (current_machine_type>=MACHINE_ID_QL_STANDARD && current_machine_type<=179)

/*
180=MK14 Standard
181-189 reservado para otros MK14
*/

#define MACHINE_IS_MK14_STANDARD (current_machine_type==MACHINE_ID_MK14_STANDARD)
#define MACHINE_IS_MK14 (current_machine_type>=MACHINE_ID_MK14_STANDARD && current_machine_type<=189)


#define CPU_IS_MOTOROLA (MACHINE_IS_QL)
#define CPU_IS_SCMP (MACHINE_IS_MK14)
#define CPU_IS_Z80 (!CPU_IS_MOTOROLA && !CPU_IS_SCMP)


extern int machine_emulate_memory_refresh;
extern int machine_emulate_memory_refresh_counter;

extern z80_bit cpu_ldir_lddr_hack_optimized;

extern z80_byte last_inves_low_ram_poke_menu;

extern void random_ram_inves(z80_byte *puntero,int longitud);

//Tipos de CPU Z80
#define TOTAL_Z80_CPU_TYPES 3
enum z80_cpu_types
{
  Z80_TYPE_GENERIC,
  Z80_TYPE_MOSTEK,
  Z80_TYPE_CMOS
};

extern enum z80_cpu_types z80_cpu_current_type;

extern char *z80_cpu_types_strings[];

//valor obtenido probando
#define MAX_EMULATE_MEMORY_REFRESH_COUNTER 1500000
#define MAX_EMULATE_MEMORY_REFRESH_LIMIT (MAX_EMULATE_MEMORY_REFRESH_COUNTER/2)


extern int zesarux_main (int main_argc,char *main_argv[]);
extern z80_bit no_cambio_parametros_maquinas_lentas;
extern z80_bit opcion_no_splash;

extern int argc;
extern char **argv;
extern int puntero_parametro;
extern int siguiente_parametro(void);

extern void siguiente_parametro_argumento(void);

extern int joystickkey_definidas;
extern void rom_load(char *romfilename);
extern void hard_reset_cpu(void);

extern z80_bit zxmmc_emulation;

extern z80_bit quickexit;

extern z80_bit azerty_keyboard;

extern z80_int ramtop_ace;

extern z80_bit allow_write_rom;

extern int z88_cpc_keymap_type;

extern z80_bit chloe_keyboard;

extern unsigned int debug_t_estados_parcial;

#define MAX_CPU_TURBO_SPEED 16

extern int cpu_turbo_speed;

extern void cpu_set_turbo_speed(void);

extern z80_bit windows_no_disable_console;

extern char *string_machines_list_description;

extern void set_menu_gui_zoom(void);


extern int exit_emulator_after_seconds;

extern int exit_emulator_after_seconds_counter;

extern z80_bit do_no_show_changelog_when_update;

extern z80_bit sdl_raw_keyboard_read;

extern z80_bit core_spectrum_uses_reduced;


extern char parameter_disablebetawarning[];

extern int total_minutes_use;

extern char macos_path_to_executable[];


#endif
